### PR TITLE
Add Activity Collector pipeline and pds-activity CLI

### PR DESCRIPTION
**Pull request by jordanpadams**
## 🗒️ Summary

Implements the full PDS EN Activity Collector pipeline across issues #61–#65. Adds a new `lasso.issues.activity` subpackage and `pds-activity` CLI that collects closed GitHub issues, merged PRs, and releases for a date range across NASA-PDS repositories, correlates them, and writes a canonical `activity.json` output file.

- `lasso.issues.activity.collector_issues` — repo discovery (explicit list or all-org, optional `--exclude-config`) + closed-issue collection; includes `parent_issue` summary (Theme/Epic) for grouping and partial-delivery signaling
- `lasso.issues.activity.collector_prs` — merged PR collection with draft exclusion
- `lasso.issues.activity.collector_releases` — GitHub Releases with tag fallback and full pagination
- `lasso.issues.activity.correlator` — body-text linkage (`closes #N`, release body refs) with transitive issue→release propagation using chronological release ordering
- `lasso.issues.activity.cli` — `pds-activity` CLI orchestrating all collectors; writes `activity.json` and `activity-validation.log`
- `docs/activity-schema.md` — full output schema reference
- 60 unit tests + 8 integration tests (marked `@pytest.mark.integration`, require `GITHUB_TOKEN`)

### Fixes applied after code review
- Wrong exception type on `gh.session.get()` calls — `requests.exceptions.RequestException` now caught alongside `github3.exceptions.GitHubException`
- `closing_release` uses chronological `published_at` order, not lexicographic tag comparison
- Releases/tags endpoints paginate past 100 via a shared `_paginate()` helper
- Process-boundary `except Exception` blocks use `_logger.exception()` for full stack traces
- Docstring corrected: collector searches for *closed* issues, not *updated*
- GitHub search 1000-result cap documented in docstring, logged as WARNING at runtime, and noted in `docs/activity-schema.md`

### 🤖 AI Assistance Disclosure
- [ ] No AI assistance used
- [ ] AI used for light assistance (e.g., suggestions, refactoring, documentation help, minor edits)
- [ ] AI used for moderate content generation (AI generated some code or logic, but the developer authored or heavily revised the majority)
- [x] AI generated substantial portions of this code

Estimated % of code influenced by AI: ~95%

> All six source modules, tests, schema documentation, and README section were generated by Claude Sonnet 4.6 via Claude Code. The developer directed the architecture, reviewed and approved the design plan, answered scoping questions (repo discovery modes, correlation depth, naming), reviewed code-review findings, and made final decisions on all design choices. Code was reviewed by a human collaborator (see inline comments) and fixes were applied before merge.

## ⚙️ Test Data and/or Report

Automated tests — 87 passed, 8 skipped (integration, require `GITHUB_TOKEN`), 0 failed:

```
tox -e py313
# py313: OK (11.53=setup[7.53]+cmd[4.00] seconds)
# congratulations :)
```

Integration tests validate the full pipeline against a live GitHub repo:
```bash
GITHUB_TOKEN=<token> pytest tests/activity/test_cli.py -m integration -v
```

Smoke test:
```bash
pds-activity --token $GITHUB_TOKEN --repos lasso-issues \
  --start-date 2026-01-01 --end-date 2026-07-11
```

## ♻️ Related Issues

Closes #61
Closes #62
Closes #63
Closes #64
Closes #65

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Documentation and PR Content
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue
- [ ] **PR Title:** The PR title is "user-friendly" clearly identifying what is being fixed or the new feature being added, that if you saw it in the Release Notes for a tool, you would be able to get the gist of what was done.

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Maintenance
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

### CODE REVIEW

**Review by nutjob4life — COMMENTED** [review]
@jordanpadams is the PR description maybe for a different branch? The diffs for this PR don't mention "evidence" at all.

| Description | Diffs |
|:------------|:------|
| New subpackage `lasso.issues.evidence` | New package is `lasso.issues.activity` |
| CLI + export command `pds-evidence` | New console script named `pds-activity` |
| Output to files `evidence.json` | Outputs to `activity.json` |

Maybe just a word swap?

**Comment by jordanpadams** [comment]
@nutjob4life it was a word swap because I didn’t particularly like “evidence”. I will update the PR and issues themselves to better reflect the changes.

**Comment by nutjob4life** [comment]
@jordanpadams cool! I'll review it with "activity" in mind 🧠

File: tests/activity/test_correlator.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        rel_early = dict(_release("v1.10.0", body_summary="#10"))
+        rel_early['published_at'] = "2026-06-01T00:00:00Z"
+        rel_late = dict(_release("v1.9.0", body_summary="#10"))
+        rel_late['published_at'] = "2026-07-01T00:00:00Z"
+
+        corr_issues, _, _, _ = correlate(issues, [pr], [rel_early, rel_late])
+
+        # Chronologically earliest is v1.10.0 (June), NOT v1.9.0 (July)
+        self.assertEqual(corr_issues[0]['closing_release'], "v1.10.0")
+
```
Comment (nutjob4life) [inline tests/activity/test_correlator.py:160]: This test? Exquisite 🧑‍🍳😗

File: docs/activity-schema.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+---
+
+## Sorting
+
+All artifact lists are sorted deterministically:
+- `issues` and `pull_requests`: by `(repo, id)` ascending
+- `releases`: by `(repo, published_at)` ascending
+- `correlation_log`: alphabetically
+
+This ensures that identical inputs always produce byte-for-byte identical output.
```
Comment (nutjob4life) [inline docs/activity-schema.md:132]: Identical? Well mostly 😁 

The output is stable _aside from_ the `generated_at` 😉

File: src/lasso/issues/activity/collector_issues.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        raise
+
+    _logger.info("Discovered %d repositories in %s", len(repo_names), org)
+    return sorted(repo_names)
+
+
+def collect_issues(gh, org: str, repos, start_date: str, end_date: str) -> list:
+    """Collect and normalize issues across the given repositories.
+
+    Searches for issues (not PRs) updated within the date range using
```
Comment (nutjob4life) [inline src/lasso/issues/activity/collector_issues.py]: Docstring says "updated within the date range" but the code (and the README) actually looks for "_closed_ within the date range".

File: src/lasso/issues/activity/collector_issues.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    # Search for closed issues in the date range
+    query = f"org:{org} is:issue is:closed closed:{start_date}..{end_date}{repo_qualifier}"
+    _logger.debug("Issue search query: %s", query)
+
+    issues: list[ActivityIssue] = []
+    seen_ids: set[int] = set()
+
+    try:
+        for search_issue in gh.search_issues(query):
```
Comment (nutjob4life) [inline src/lasso/issues/activity/collector_issues.py:145]: GitHub puts a 1000 result limit on `search_issues`. Might be worth a warning in the logs or docs?

**Review by nutjob4life — COMMENTED** [review]
Delta: ✓
Tox: 👎 
Approval: contigent on `tox` disposition

- Great and clear modular split between collectors, schema, correlator, and CLI.
- The correlator tests are a thing of beauty.
- There have been a number of PRs for `lasso-issues` recently but none have addressed the `tox` problems. Is the linting _supposed_ to fail in this repository?
- I had some incredibly minor interspersed comments in the diffs, all of each are easily ignored!

Tox details:
```
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing examples/b18_schedule_prep_spreadsheet.csv

Check that executables have shebangs.................(no files to check)Skipped
Check for merge conflicts................................................Passed
Debug Statements (Python)................................................Passed
Check Yaml...............................................................Passed
Reorder python imports...................................................Failed
- hook id: reorder-python-imports
- exit code: 1
- files were modified by this hook

Reordering imports in src/lasso/issues/activity/cli.py
Reordering imports in tests/activity/test_collector_releases.py
Reordering imports in src/lasso/issues/activity/collector_prs.py
Reordering imports in src/lasso/issues/activity/collector_issues.py
Reordering imports in src/lasso/issues/activity/collector_releases.py

flake8...................................................................Failed
- hook id: flake8
- exit code: 1

src/lasso/issues/activity/cli.py:125:9: F541 f-string is missing placeholders
src/lasso/issues/activity/cli.py:129:9: F541 f-string is missing placeholders
src/lasso/issues/activity/cli.py:130:9: F541 f-string is missing placeholders
src/lasso/issues/activity/cli.py:135:9: F541 f-string is missing placeholders
src/lasso/issues/issues/issues.py:29:1: E303 too many blank lines (3)
src/lasso/issues/issues/utils.py:102:42: E741 ambiguous variable name 'l'
src/lasso/issues/issues/utils.py:107:27: E741 ambiguous variable name 'l'
src/lasso/issues/issues/utils.py:111:52: E741 ambiguous variable name 'l'

Detect secrets...........................................................Passed
lint: exit 1 (2.75 seconds) /Users/kelly/Documents/Clients/JPL/PDS/Development/nasa-pds/lasso-issues> python -m pre_commit run --color=always --all pid=60847
  py313: OK (27.65=setup[16.19]+cmd[11.47] seconds)
  docs: OK (13.85=setup[12.54]+cmd[1.30] seconds)
  lint: FAIL code 1 (4.09=setup[1.34]+cmd[2.75] seconds)
  evaluation failed :( (45.67 seconds)
```

File: tests/activity/test_correlator.py (hunk 1 above)
Comment (jordanpadams) [reply tests/activity/test_correlator.py:160]: Ha, glad it landed! That test case was fun to write — needed to make sure we weren't accidentally re-introducing the lexicographic semver bug in a future refactor.

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

File: docs/activity-schema.md (hunk 2 above)
Comment (jordanpadams) [reply docs/activity-schema.md:132]: Fair catch 😄 — clarified in the schema doc: `generated_at` is intentionally a run timestamp and is the one field that will differ between runs on the same inputs. Added a note to the Known Limitations / Sorting section to make that explicit.

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

File: src/lasso/issues/activity/collector_issues.py (hunk 3 above)
Comment (jordanpadams) [reply src/lasso/issues/activity/collector_issues.py]: Good catch — fixed in the latest commit. The docstring now reads "Collect and normalize issues **closed** within the date range" and explicitly documents the 1000-result cap with mitigation advice.

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

File: src/lasso/issues/activity/collector_issues.py (hunk 4 above)
Comment (jordanpadams) [reply src/lasso/issues/activity/collector_issues.py:145]: Great point — addressed in the latest commit:
- Docstring on `collect_issues` now documents the 1000-result cap and suggests narrowing the date range or using `--repos` as mitigations
- A `WARNING` is logged at runtime when the result count hits 1000
- Added a "Known Limitations" section to `docs/activity-schema.md` calling this out explicitly

The same cap technically applies to the PR collector too, so the same warning logic applies there as well — filed a follow-on to add that.

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

**Comment by jordanpadams** [comment]
@nutjob4life made some updates based upon your feedback, and linter is 🟢 .

**Review by nutjob4life — APPROVED** [review]
> @nutjob4life made some updates based upon your feedback, and linter is 🟢 .

Woo-hoo! Thanks for putting up with all that @jordanpadams.

My own linting is still failing for some reason? I will look at that time permitting later. For now, approval is ✅.

```console
bash-5.3$ cd /tmp
bash-5.3$ git clone --quiet --branch feat/evidence-collector git@github.com:NASA-PDS/lasso-issues.git
bash-5.3$ cd lasso-issues
bash-5.3$ pre-commit install
pre-commit install --hook-type pre-push
pre-commit install --hook-type prepare-commit-msg
pre-commit install --hook-type commit-msg
pre-commit installed at .git/hooks/pre-commit
pre-commit installed at .git/hooks/pre-push
pre-commit installed at .git/hooks/prepare-commit-msg
pre-commit installed at .git/hooks/commit-msg
bash-5.3$ python3.13 -m venv .venv
bash-5.3$ . .venv/bin/activate
(.venv) bash-5.3$ pip install --quiet --editable '.[dev]'

[notice] A new release of pip is available: 25.2 -> 26.1.2
[notice] To update, run: pip install --upgrade pip
(.venv) bash-5.3$ tox -e lint
lint: install_deps> python -I -m pip install pre-commit
lint: commands[0]> python -m pre_commit run --color=always --all
[WARNING] hook id `tests` uses deprecated stage names (push) which will be removed in a future version.  run: `pre-commit migrate-config` to automatically fix this.
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing examples/b18_schedule_prep_spreadsheet.csv

Check that executables have shebangs.................(no files to check)Skipped
Check for merge conflicts................................................Passed
Debug Statements (Python)................................................Passed
Check Yaml...............................................................Passed
Reorder python imports...................................................Failed
- hook id: reorder-python-imports
- exit code: 1
- files were modified by this hook

Reordering imports in src/lasso/issues/activity/cli.py
Reordering imports in tests/activity/test_collector_releases.py
Reordering imports in src/lasso/issues/activity/collector_prs.py
Reordering imports in src/lasso/issues/activity/collector_issues.py
Reordering imports in src/lasso/issues/activity/collector_releases.py

flake8...................................................................Failed
- hook id: flake8
- exit code: 1

src/lasso/issues/activity/cli.py:125:9: F541 f-string is missing placeholders
src/lasso/issues/activity/cli.py:129:9: F541 f-string is missing placeholders
src/lasso/issues/activity/cli.py:130:9: F541 f-string is missing placeholders
src/lasso/issues/activity/cli.py:135:9: F541 f-string is missing placeholders
src/lasso/issues/issues/issues.py:29:1: E303 too many blank lines (3)
src/lasso/issues/issues/utils.py:102:42: E741 ambiguous variable name 'l'
src/lasso/issues/issues/utils.py:107:27: E741 ambiguous variable name 'l'
src/lasso/issues/issues/utils.py:111:52: E741 ambiguous variable name 'l'

Detect secrets...........................................................Passed
lint: exit 1 (2.55 seconds) /private/tmp/lasso-issues> python -m pre_commit run --color=always --all pid=65208
  lint: FAIL code 1 (3.82=setup[1.27]+cmd[2.55] seconds)
  evaluation failed :( (3.89 seconds)
```
